### PR TITLE
Keyboard Navigation enhancements

### LINKED ISSUE #6747: Keyboard navigation service enhancements
## Description  
This issue is part of the general kb navigation enhancements that we plan to work on next sprints. The following problems should be covered:
- [x] `Keydown` handler to be on Grid level
- [x] Enhance Grid navigation service - Currently, we have four navigation services for the Grid, HierarchicalGrid, TreeGrid, and MRL functionality.
- [x] Remove legacy code whenever possible

### LINKED ISSUE #6835: Errors are thrown when scrolling the grid with touch interaction
## Description  
Errors are thrown when scrolling the grid with touch interaction.

 * igniteui-angular version: 9.0
 * browser: NA

## Steps to reproduce  

1. Open https://www.infragistics.com/products/ignite-ui-angular/angular/components/grid/live_data.html or https://www.infragistics.com/products/ignite-ui-angular/angular/components/treegrid/live_data.html sample.
2. Collapse some rows.
3. Scroll using touch interaction.

## Result  
Errors are thrown:
- Uncaught DOMException: Failed to execute 'insertBefore' on 'Node': The node to be removed is no longer a child of this node. Perhaps it was moved in a 'blur' event handler?
- ERROR TypeError: Cannot read property 'context' of undefined

## Expected result  
No exceptions should be thrown.

### LINKED ISSUE #6882: KB Nav. MRL
Update Keyboard Navigation for MRL

### PR DESCRIPTION
**Pull request by ddincheva**
Closes #6747 
Closes #6882 (KB Nav. MRL)
Closes #6835

### Additional information (check all that apply):
 - [ ] Bug fix
 - [ ] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [ ] All relevant tags have been applied to this PR
 - [ ] This PR includes unit tests covering all the new code
 - [ ] This PR includes API docs for newly added methods/properties
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [ ] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### Issues:

- [x] 1. When open a cell in edit mode and press Tab key the events onCellEdit and onCellEditEnter are called 2 times

- [x] 2. MRL: Arrow Left key does not work correct. Create a grid similar to the ColumnLayoutTestComponent. Click on a cell in column ContactName and press arrowLeft. First cell is selected instead of previous.

- [x] 3. Row Editing: when a custom row editing template and press tab, the next row row cell is selected. Can be reproduced in the gridRowEdit sample with clicking Custom template.

- [x] 4. Row editing: when the grid is scrolled and press tab on apply button not correct cell is selected:
rowEditing

- [x] 5. When focus the grid with the tab key the event onSelection should be fired

- [x] 6. When click on a Masterdetail and press Ctrl+arrow down/up the grid is not scrolled (this behavior stays as it is (The spec should be updated)) (related issue: [7048](https://github.com/IgniteUI/igniteui-angular/issues/7048))

- [x] 7. When navigate with tab or shift+tab the grid should be scrolled to the first cell. Open dev sample gridSummary. scroll down and click on a cell. Press tab key. Press Shift + tab. the first cell is not focused.
- [x] 8. MRL: Navigating with Ctr+ Arrow left doe not navigate to the correct cell. 
Steps to reproduce:
1. Open dev sample gridMRL
2. Hide column ID
3. Click on a cell in column Fax
4. press Ctrl + arrow Left
5. cell in column CompanyName is selected, but the cell in column ContactName should be sected
see unit tests:
- should navigate to the first cell from the layout by pressing Ctrl + Arrow Left and Right key
- should navigate to the first cell from the layout by pressing Ctrl + Arrow Right and Left key
- [x]  9. MRL: Navigating with arrow Up/Down does not navigate to the correct cell. see unit tests:
-  should navigate using Arrow Down through cell with smaller colStart and bigger colEnd
-  should navigate using Arrow Down through cell with same colStart but bigger colEnd
- should navigate using Arrow Up through cell with smaller colStart and bigger colEnd
- should navigate down to next row inside pinned area when pinning second block in template
- should navigate down to next row inside unpinned area when pinning second block in template
- should navigate up to next row inside unpinned area when pinning second block in template

- [x]  10. If there is no active node navigateTo method returns error.
See unit test:
- navigateTo method should work in multi-row layout grid

- [x] 11. Row Editing: when there is a custom row editing template with buttons using IgxRowEditTabStop directives the navigation never goes through them. (Also reproducible in master)[[7049](https://github.com/IgniteUI/igniteui-angular/issues/7049)]

- [x] 12. In Edge when press Esc key on a cell in edit mode it is not possible to continue with keyboard navigation.

- [x] 13. When from the last row press Shift+Tab from edited cell  the header is focused.
Steps to reproduce:
1. Open dev sample gridCellEditing
2. Scroll to the button and in the last row double click on the cell in column ProductID
3. Press shift+tab
4. The header is focused, but the previous editable cell should be opened in edit mode.

- [ ] 14. Navigation is broken in Chrome when in zoom is applied in the display setting. See [coment](https://github.com/IgniteUI/igniteui-angular/pull/6910#issuecomment-603709560)

- [x] 15. Navigation between summaries tbody and headers is not working in some cases:
- shift + tab from summary to tbody when there are/is a focusable element(s) brokes the behaviour and does not select the right cell
- shift + tab from summary to tbody without focusable element(s) selects the right cell but does not move the selectable style.
- When navigating from tbody to header errors are thrown due to incorrect focus application. (mrl sample)
- When navigating from outside the grid to the headers section nothing happens and lacks an ability to interact with it.

- [x] 16 When grid is scrolled horizontal it is not possible to resize a column. 
Steps to reproduce:
1. Open dev sample gridColumnResizing
2. Scroll horizontal and start resize column City
3. Th scroll is moved to the beginning and the column cannot be resized

### CODE REVIEW

**Comment by dkamburov** [comment]
closes #6835

**Comment by dkamburov** [comment]
Here's two things I noticed. I know they could be considered as edge-case scenarios. And I'm not saying they must be resolved for the PR to be merged, but some users may notice them too:
1. When a cell width is bigger than display container width there is flickering:
flickering
2. If there is zoom applied, ctrl + home/end is unable to scroll properly to the target cell. Similar behavior can be reproduced holding down ArrowUp or ArrowDown:
ctrl+home

**Comment by nrobakova** [comment]
Hi I am adding the issues which I found:
1. When there is selected cells range and press Space key to select a row the cell selection is cleared
2. When Navigate from cell to group row with Shift key and then from group row navigate without shift the previous selection is not cleared
3. When a cell is selected and the grid is srolled to top navigating with PgUp scrolls the page
4. When a cell is opened in edit press tab to got to the next editable, press Esc, the current cell is not selected.
5. When all group rows are collapsed and click on a group row and press Ctrl+End, after that it is not possible to scroll the grid

File: projects/igniteui-angular/src/lib/grids/summaries/summary-cell.component.ts [hunk 1]
```diff
@@ -40,25 +44,20 @@ export class IgxSummaryCellComponent {
...
+    public get attrCellID() {
+        return `${this.grid.id}_${this.rowIndex}_${ this.visibleColumnIndex}`;
     }
 
-    @HostListener('blur')
-    public onBlur() {
-        this.focused = false;
+    @HostListener('pointerdown')
+    public activate() {
+        this.grid.navigation.activeNode = {row: this.rowIndex, column: this.visibleColumnIndex};
```
Comment (zdrawku) [inline projects/igniteui-angular/src/lib/grids/summaries/summary-cell.component.ts]: Can we think of a way to use interface or class for the activeNode objects. This object assignment is used very often in the code

**Comment by dkamburov** [comment]
Not sure if this is intended, but it seems we are introducing a behavioral change with cell editing tab navigation. Before these changes navigating to a new row causes the updating to be closed, now the editing continues on  the new row:
celledit
It may sound like a minor thing, but if we want it that we, there might be issues with groupby, master-detail, hierarchical grid etc.

**Comment by ddincheva** [comment]
@dkamburov we are introducing the following behavioral change 
image
The rest of the changes can be found in the [spec file](https://github.com/IgniteUI/igniteui-angular/wiki/igxGrid-Specification#keyboard-navigation).
If there is an issues with the described behavior please describe them in a comment to this PR.

**Pull request merged**